nifi-nar-bundles/nifi-mqtt-bundle/nifi-mqtt-processors/src/test/java/org/apache/nifi/processors/mqtt/TestPublishMQTT.java

```diff
@@ -17,52 +17,368 @@
+        verifyPublishedMessage(testInput.get(2).toString().getBytes(), 2, false);
+        verifyNoMorePublished();
+
+        final List<MockFlowFile> flowFiles = testRunner.getFlowFilesForRelationship(REL_SUCCESS);
+        assertEquals(1, flowFiles.size());
+
+        final MockFlowFile successfulFlowFile = flowFiles.get(0);
+        assertNull(successfulFlowFile.getAttribute(publishFailedIndexAttributeName),
+                publishFailedIndexAttributeName + " is expected to be removed after all remaining records have been published successfully.");
+    }
+
+    private void verifyPublishedMessage(byte[] payload, int qos, boolean retain) {
+        final Pair<String, StandardMqttMessage> lastPublished = mqttTestClient.getLastPublished();
+        final String lastPublishedTopic = lastPublished.getLeft();
+        final StandardMqttMessage lastPublishedMessage = lastPublished.getRight();
         assertEquals(Arrays.toString(payload), Arrays.toString(lastPublishedMessage.getPayload()));
         assertEquals(qos, lastPublishedMessage.getQos());
         assertEquals(retain, lastPublishedMessage.isRetained());
-        assertEquals(topic, lastPublishedTopic);
+        assertEquals(TOPIC, lastPublishedTopic);
     }
 
-    private MqttTestClient mqttTestClient;
+    private void verifyNoMorePublished() {
+        assertNull(mqttTestClient.getLastPublished(), "TestClient's queue should be empty.");
+    }
+
+    private ProvenanceEventRecord assertProvenanceEvent() {
+        final List<ProvenanceEventRecord> provenanceEvents = testRunner.getProvenanceEvents();
+        assertNotNull(provenanceEvents);
+        assertEquals(1, provenanceEvents.size());
 
-    public class UnitTestablePublishMqtt extends PublishMQTT {
+        final ProvenanceEventRecord event = provenanceEvents.get(0);
+        assertEquals(ProvenanceEventType.SEND, event.getEventType());
 
-        public UnitTestablePublishMqtt(){
-            super();
-        }
+        return event;
+    }
 
-        @Override
-        protected MqttClient createMqttClient() throws MqttException {
-            mqttTestClient = new MqttTestClient(MqttTestClient.ConnectType.Publisher);
-            return mqttTestClient;
-        }
+    private void assertProvenanceEvent(String expectedDetails) {
+        final ProvenanceEventRecord event = assertProvenanceEvent();
+        assertEquals(expectedDetails, event.getDetails());
     }
 
-    @BeforeEach
-    public void init() {
-        UnitTestablePublishMqtt proc = new UnitTestablePublishMqtt();
-        testRunner = TestRunners.newTestRunner(proc);
-        testRunner.setProperty(PublishMQTT.PROP_BROKER_URI, "tcp://localhost:1883");
-        testRunner.setProperty(PublishMQTT.PROP_RETAIN, "false");
-        topic = "testTopic";
-        testRunner.setProperty(PublishMQTT.PROP_TOPIC, topic);
+    private static ArrayNode createTestJsonInput() {
+        final ObjectMapper mapper = new ObjectMapper();
```

Comment:
I think it is a best practice to initialize `ObjectMapper` once and store it in a `static` field because it seems to be a heavy-weight object but thread-safe.

It can be modified in a future commit. Will go ahead with merging this PR.